c10/core/ScalarType.cpp

```diff
@@ -1,23 +1,48 @@
 #include <c10/core/ScalarType.h>
+#include <array>
 
 namespace c10 {
 
+namespace {
+
+constexpr auto u1 = ScalarType::Byte;
+constexpr auto i1 = ScalarType::Char;
+constexpr auto i2 = ScalarType::Short;
+constexpr auto i4 = ScalarType::Int;
+constexpr auto i8 = ScalarType::Long;
+constexpr auto f2 = ScalarType::Half;
+constexpr auto f4 = ScalarType::Float;
+constexpr auto f8 = ScalarType::Double;
+constexpr auto c2 = ScalarType::ComplexHalf;
+constexpr auto c4 = ScalarType::ComplexFloat;
+constexpr auto c8 = ScalarType::ComplexDouble;
+constexpr auto b1 = ScalarType::Bool;
+constexpr auto bf = ScalarType::BFloat16;
+constexpr auto ud = ScalarType::Undefined;
+
+constexpr int64_t NUM_PROMOTE_TYPES = 20;
```

Comment:
Couldn't you use array_of to construct the array below and then set this constant by exploiting the fact the std::array::size is a constexpr? Wouldn't this make it less error prone to forgetting to update this in the future potentially?

Reply:
Oh I see, it's not a stdlib thing but there's a helper hanging out in an inconvenient place. I guess I'll move it over.